include/sys/sa_impl.h

```diff
@@ -177,8 +177,11 @@ typedef struct sa_hdr_phys {
 	 *
 	 */
 	uint16_t sa_layout_info;
-	uint16_t sa_lengths[1];	/* optional sizes for variable length attrs */
-	/* ... Data follows the lengths.  */
+	uint16_t sa_lengths[VARLEN_ARRAY_IDX];
```

Comment:
Changing the size of this member seem like it could cause compatibility problems since it's an on-disk struct.  There are few places like `SA_HDR_SIZE_MATCH_LAYOUT` which call `sizeof (sa_hdr_phys_t)` and we don't want this returning different sizes depending on how `VARLEN_ARRAY_IDX` is defined.

Reply:
For this particular one:
```c
 #define SA_HDR_SIZE_MATCH_LAYOUT(hdr, tb) \
         (SA_HDR_SIZE(hdr) == (sizeof (sa_hdr_phys_t) + \
         (tb->lot_var_sizes > 1 ? P2ROUNDUP((tb->lot_var_sizes - 1) * \
         sizeof (uint16_t), 8) : 0)))
```

I'm not 100% certain the details of the `tb` argument, but taking some context clues it looks like the trinary operation is attempting to extend the offset by `tb->lot_var_sizes - 1` if the count is greater than `1`, and by `0` otherwise - which I suspect is trying to adjust for the fact that the `sa_lengths` array is already `1` long? AFAICT, this is also only used in one place in `module/zfs/sa.c` ... so would it make sense to adjust this macro to use a simpler calculation for the case where `VARLEN_ARRAY_IDX` is `""`?

Reply:
however data follows after this. and thus also follows on disk. how is that accounted for? surely things are expected to be 1 or more after here.